File: eggfinder/board.py
import random

from .constants import CellType, GameConfig
from .exceptions import InvalidBoardError


class Board:
    """Board state and operations."""

    def __init__(
        self,
        width: int = GameConfig.DEFAULT_BOARD_WIDTH,
        height: int = GameConfig.DEFAULT_BOARD_HEIGHT,
        egg_count: int = GameConfig.DEFAULT_EGG_COUNT,
    ):
        if width <= 0 or height <= 0:
            raise InvalidBoardError(f"Board dimensions must be positive, got {width}x{height}")

        max_eggs = width * height
        if egg_count < 0 or egg_count > max_eggs:
            raise InvalidBoardError(f"Egg count must be between 0 and {max_eggs}, got {egg_count}")

        self.width = width
        self.height = height
        self.egg_count = egg_count
        self.cells: list[list[int]] = []
        self.revealed: list[list[bool]] = []
        self.eggs: set[tuple[int, int]] = set()

        self._initialize()
        self._place_eggs()
        self._calculate_numbers()

    def _initialize(self) -> None:
        self.cells = [[0 for _ in range(self.width)] for _ in range(self.height)]
        self.revealed = [[False for _ in range(self.width)] for _ in range(self.height)]

    def _place_eggs(self) -> None:
        placed = 0
        while placed < self.egg_count:
            row = random.randint(0, self.height - 1)
            col = random.randint(0, self.width - 1)

            if (row, col) not in self.eggs:
                self.eggs.add((row, col))
                self.cells[row][col] = CellType.EGG
                placed += 1

    def _calculate_numbers(self) -> None:
        """Calculate adjacent egg counts for each cell."""
        for row in range(self.height):
            for col in range(self.width):
                if self.cells[row][col] != CellType.EGG:
                    self.cells[row][col] = self._count_adjacent_eggs(row, col)

    def _count_adjacent_eggs(self, row: int, col: int) -> int:
        """Count eggs adjacent to the given position."""
        return sum(1 for nr, nc in self.get_neighbors(row, col) if self.is_egg(nr, nc))

    def is_valid_position(self, row: int, col: int) -> bool:
        return 0 <= row < self.height and 0 <= col < self.width

    def get_cell(self, row: int, col: int) -> int | None:
        if not self.is_valid_position(row, col):
            return None
        return self.cells[row][col]

    def is_egg(self, row: int, col: int) -> bool:
        return (row, col) in self.eggs

    def is_revealed(self, row: int, col: int) -> bool:
        if not self.is_valid_position(row, col):
            return False
        return self.revealed[row][col]

    def get_neighbors(self, row: int, col: int) -> list[tuple[int, int]]:
        neighbors = []
        for dr in [-1, 0, 1]:
            for dc in [-1, 0, 1]:
                if dr == 0 and dc == 0:
                    continue
                nr, nc = row + dr, col + dc
                if self.is_valid_position(nr, nc):
                    neighbors.append((nr, nc))
        return neighbors

File: eggfinder-rs/src/board.rs
//! Board state and operations.

use std::collections::HashSet;
use std::fmt;

use rand::Rng;

use crate::constants::CellType;

/// Coordinate type alias (row, col).
pub type Coordinate = (usize, usize);

/// Errors that can occur when creating or manipulating a board.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum BoardError {
    InvalidDimensions { width: usize, height: usize },
    TooManyEggs { egg_count: usize, max_cells: usize },
}

impl fmt::Display for BoardError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            BoardError::InvalidDimensions { width, height } => {
                write!(
                    f,
                    "Board dimensions must be positive, got {}x{}",
                    width, height
                )
            }
            BoardError::TooManyEggs {
                egg_count,
                max_cells,
            } => {
                write!(
                    f,
                    "Egg count {} exceeds board size {}",
                    egg_count, max_cells
                )
            }
        }
    }
}

impl std::error::Error for BoardError {}

/// Board state containing the grid, revealed cells, and egg locations.
pub struct Board {
    pub width: usize,
    pub height: usize,
    pub egg_count: usize,
    /// Cell values: -1 = egg, 0 = empty, 1-8 = adjacent egg count.
    pub cells: Vec<Vec<i8>>,
    pub revealed: Vec<Vec<bool>>,
    pub eggs: HashSet<Coordinate>,
}

impl Board {
    /// Create a new board with random egg placement.
    ///
    /// # Errors
    /// Returns `BoardError::InvalidDimensions` if width or height is 0.
    /// Returns `BoardError::TooManyEggs` if egg_count exceeds board size.
    pub fn new(width: usize, height: usize, egg_count: usize) -> Result<Self, BoardError> {
        if width == 0 || height == 0 {
            return Err(BoardError::InvalidDimensions { width, height });
        }

        let max_cells = width * height;
        if egg_count > max_cells {
            return Err(BoardError::TooManyEggs {
                egg_count,
                max_cells,
            });
        }

        let mut board = Board {
            width,
            height,
            egg_count,
            cells: vec![vec![0; width]; height],
            revealed: vec![vec![false; width]; height],
            eggs: HashSet::new(),
        };

        board.place_eggs();
        board.calculate_numbers();
        Ok(board)
    }

    /// Place eggs randomly on the board.
    fn place_eggs(&mut self) {
        let mut rng = rand::rng();
        let mut placed = 0;

        while placed < self.egg_count {
            let row = rng.random_range(0..self.height);
            let col = rng.random_range(0..self.width);

            if !self.eggs.contains(&(row, col)) {
                self.eggs.insert((row, col));
                self.cells[row][col] = CellType::Egg.value();
                placed += 1;
            }
        }
    }

    /// Calculate adjacent egg counts for each non-egg cell.
    fn calculate_numbers(&mut self) {
        for row in 0..self.height {
            for col in 0..self.width {
                if self.cells[row][col] != CellType::Egg.value() {
                    self.cells[row][col] = self.count_adjacent_eggs(row, col);
                }
            }
        }
    }

    fn count_adjacent_eggs(&self, row: usize, col: usize) -> i8 {
        self.get_neighbors(row, col)
            .iter()
            .filter(|&&(r, c)| self.is_egg(r, c))
            .count() as i8
    }

    pub fn is_valid_position(&self, row: usize, col: usize) -> bool {
        row < self.height && col < self.width
    }

    /// Get the cell value at a position. Returns None if out of bounds.
    pub fn get_cell(&self, row: usize, col: usize) -> Option<i8> {
        if self.is_valid_position(row, col) {
            Some(self.cells[row][col])
        } else {
            None
        }
    }

    pub fn is_egg(&self, row: usize, col: usize) -> bool {
        self.eggs.contains(&(row, col))
    }

    pub fn is_revealed(&self, row: usize, col: usize) -> bool {
        if self.is_valid_position(row, col) {
            self.revealed[row][col]
        } else {
            false
        }
    }

    /// Get all valid neighboring positions (8-directional).
    pub fn get_neighbors(&self, row: usize, col: usize) -> Vec<Coordinate> {
        let mut neighbors = Vec::new();
        let directions: [(i32, i32); 8] = [
            (-1, -1),
            (-1, 0),
            (-1, 1),
            (0, -1),
            (0, 1),
            (1, -1),
            (1, 0),
            (1, 1),
        ];

        for (dr, dc) in directions {
            let nr = row as i32 + dr;
            let nc = col as i32 + dc;

            if nr >= 0 && nc >= 0 {
                let nr = nr as usize;
                let nc = nc as usize;
                if self.is_valid_position(nr, nc) {
                    neighbors.push((nr, nc));
                }
            }
        }

        neighbors
    }

    /// Reveal a cell. Returns true if the cell was newly revealed.
    pub fn reveal(&mut self, row: usize, col: usize) -> bool {
        if self.is_valid_position(row, col) && !self.revealed[row][col] {
            self.revealed[row][col] = true;
            true
        } else {
            false
        }
    }
}
